src/java.base/share/classes/java/util/stream/Gatherers.java

```diff
@@ -350,86 +351,101 @@ public static <T, R> Gatherer<T,?,R> mapConcurrent(
-                Thread.startVirtualThread(task);
+                // Flush at least 1 element if we're at capacity
+                return flush(wip.size() < maxConcurrency ? 0 : 1, downstream);
             }
 
-            final boolean flush(long atLeastN,
-                                Downstream<? super R> downstream) {
-                boolean proceed = !downstream.isRejecting();
-                boolean interrupted = false;
+            boolean flush(long atLeastN, Downstream<? super R> downstream) {
+                boolean success = false, interrupted = false;
                 try {
-                    Future<R> current;
+                    boolean proceed = !downstream.isRejecting();
+                    MapConcurrentTask current;
                     while (proceed
-                            && (current = window.peek()) != null
-                                && (current.isDone() || atLeastN > 0)) {
-                        proceed &= downstream.push(current.get());
+                        && (current = wip.peekFirst()) != null
+                        && (current.isDone() || atLeastN > 0)) {
+                        R result;
+
+                        // Ensure that the task is done before proceeding
+                        for(;;) {
+                            try {
+                                result = current.get();
+                                break;
+                            } catch (InterruptedException ie) {
+                                interrupted = true; // ignore for now, and restore later
+                            }
+                        }
+
+                        proceed &= downstream.push(result);
                         atLeastN -= 1;
 
-                        var correctRemoval = window.pop() == current;
+                        final var correctRemoval = wip.pollFirst() == current;
                         assert correctRemoval;
                     }
-                } catch(InterruptedException ie) {
-                    proceed = false;
-                    interrupted = true;
+                    return (success = proceed); // Ensure that cleanup occurs if needed
                 } catch (ExecutionException e) {
-                    proceed = false; // Ensure cleanup
                     final var cause = e.getCause();
                     throw (cause instanceof RuntimeException re)
                               ? re
                               : new RuntimeException(cause == null ? e : cause);
                 } finally {
-                    // Clean up
-                    if (!proceed) {
-                        Future<R> next;
-                        while ((next = window.pollFirst()) != null) {
-                            next.cancel(true);
+                    // Clean up work-in-progress
+                    if (!success && !wip.isEmpty()) {
+                        // First signal cancellation for all tasks in progress
+                        for(var task : wip)
```

Comment:
Minor formating nit is that you probably want a space in "for(", there are a few more of these in the patch.

Reply:
Fixed!